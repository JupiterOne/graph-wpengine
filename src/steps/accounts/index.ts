import {
  createDirectRelationship,
  Entity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';

import { IntegrationConfig } from '../../config';
import { createAPIClient } from '../../client';
import {
  Entities,
  Relationships,
  IntegrationSteps,
  USER_ENTITY_KEY,
} from '../constants';
import { createAccountEntity } from './converters';

export async function fetchAccounts({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  const userEntity = (await jobState.getData(USER_ENTITY_KEY)) as Entity;

  await apiClient.iterateAccounts(async (account) => {
    const accountEntity = createAccountEntity(account);

    await jobState.addEntity(accountEntity);

    if (userEntity) {
      await jobState.addRelationship(
        createDirectRelationship({
          _class: RelationshipClass.HAS,
          from: userEntity,
          to: accountEntity,
        }),
      );
    }
  });
}

export const accountSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: IntegrationSteps.ACCOUNTS,
    name: 'Fetch Accounts',
    entities: [Entities.ACCOUNT],
    relationships: [Relationships.USER_HAS_ACCOUNT],
    dependsOn: [IntegrationSteps.USER],
    executionHandler: fetchAccounts,
  },
];
