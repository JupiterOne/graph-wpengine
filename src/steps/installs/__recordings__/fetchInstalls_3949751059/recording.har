{
  "log": {
    "_recordingName": "fetchInstalls",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "4.3.0"
    },
    "entries": [
      {
        "_id": "b5a41916ddb8fd6f9a367486771ee186",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "api.wpengineapi.com"
            }
          ],
          "headersSize": 352,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "limit",
              "value": "10"
            },
            {
              "name": "offset",
              "value": "0"
            }
          ],
          "url": "https://api.wpengineapi.com/v1/accounts?limit=10&offset=0"
        },
        "response": {
          "bodySize": 112,
          "content": {
            "mimeType": "application/json",
            "size": 112,
            "text": "{\"previous\":null,\"next\":null,\"count\":1,\"results\":[{\"id\":\"77815c5d-e25c-41a3-ba2d-29f8ceb82ac1\",\"name\":\"j1inc\"}]}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "content-length",
              "value": "112"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "date",
              "value": "Mon, 05 Jul 2021 11:52:23 GMT"
            },
            {
              "name": "x-amzn-requestid",
              "value": "7d137e28-49a1-44f1-b654-f718760cd04b"
            },
            {
              "name": "access-control-allow-origin",
              "value": "*"
            },
            {
              "name": "access-control-allow-headers",
              "value": "Content-Type,X-Amz-Date,Authorization,X-Api-Key"
            },
            {
              "name": "x-amz-apigw-id",
              "value": "B_rSkFZaIAMFr6w="
            },
            {
              "name": "access-control-allow-methods",
              "value": "*"
            },
            {
              "name": "x-amzn-trace-id",
              "value": "Root=1-60e2f276-000b541f55287a7559cc98e6"
            },
            {
              "name": "x-cache",
              "value": "Miss from cloudfront"
            },
            {
              "name": "via",
              "value": "1.1 35ca6aa74c5646f0333fa6da149920ce.cloudfront.net (CloudFront)"
            },
            {
              "name": "x-amz-cf-pop",
              "value": "MNL50-C1"
            },
            {
              "name": "x-amz-cf-id",
              "value": "f32KzdQu60OR0n45om99wx9SLJT_jhfBRn-U3j49RqVcS1MUd_BhEg=="
            }
          ],
          "headersSize": 601,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-07-05T11:52:22.050Z",
        "time": 1253,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 1253
        }
      },
      {
        "_id": "97123c15bf93205301118bdc565d93ea",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "api.wpengineapi.com"
            }
          ],
          "headersSize": 349,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "limit",
              "value": "10"
            },
            {
              "name": "offset",
              "value": "0"
            }
          ],
          "url": "https://api.wpengineapi.com/v1/sites?limit=10&offset=0"
        },
        "response": {
          "bodySize": 354,
          "content": {
            "mimeType": "application/json",
            "size": 354,
            "text": "{\"previous\":null,\"next\":null,\"count\":1,\"results\":[{\"id\":\"4d91f2dd-9743-421e-9bcf-20dcf0a195b2\",\"name\":\"jupiterone\",\"account\":{\"id\":\"77815c5d-e25c-41a3-ba2d-29f8ceb82ac1\"},\"installs\":[{\"id\":\"30c47931-a91d-468a-a5dc-b9711f06c60b\",\"name\":\"jupiterone\",\"environment\":\"production\",\"cname\":\"jupiterone.wpengine.com\",\"php_version\":\"7.3\",\"is_multisite\":false}]}]}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "content-length",
              "value": "354"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "date",
              "value": "Mon, 05 Jul 2021 11:52:23 GMT"
            },
            {
              "name": "x-amzn-requestid",
              "value": "a34293ef-81bf-4e4a-9991-cf94437f7b7f"
            },
            {
              "name": "access-control-allow-origin",
              "value": "*"
            },
            {
              "name": "access-control-allow-headers",
              "value": "Content-Type,X-Amz-Date,Authorization,X-Api-Key"
            },
            {
              "name": "x-amz-apigw-id",
              "value": "B_rSrEsKoAMFQ_A="
            },
            {
              "name": "access-control-allow-methods",
              "value": "*"
            },
            {
              "name": "x-amzn-trace-id",
              "value": "Root=1-60e2f277-3d90425931a5204d2355f97a"
            },
            {
              "name": "x-cache",
              "value": "Miss from cloudfront"
            },
            {
              "name": "via",
              "value": "1.1 fec87288bd626f32b2e1177a1ee70226.cloudfront.net (CloudFront)"
            },
            {
              "name": "x-amz-cf-pop",
              "value": "MNL50-C1"
            },
            {
              "name": "x-amz-cf-id",
              "value": "v4AHxG9ZXI5draCzzYGu6_HP8PTT5px3NGbXI5XmCwVbAeNluIJP0A=="
            }
          ],
          "headersSize": 601,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-07-05T11:52:23.343Z",
        "time": 648,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 648
        }
      },
      {
        "_id": "7c0782aad0b9678323207b7733ae2209",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "api.wpengineapi.com"
            }
          ],
          "headersSize": 352,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "limit",
              "value": "10"
            },
            {
              "name": "offset",
              "value": "0"
            }
          ],
          "url": "https://api.wpengineapi.com/v1/installs?limit=10&offset=0"
        },
        "response": {
          "bodySize": 398,
          "content": {
            "mimeType": "application/json",
            "size": 398,
            "text": "{\"previous\":null,\"next\":null,\"count\":1,\"results\":[{\"id\":\"30c47931-a91d-468a-a5dc-b9711f06c60b\",\"name\":\"jupiterone\",\"cname\":\"jupiterone.wpengine.com\",\"stable_ips\":null,\"status\":\"active\",\"account\":{\"id\":\"77815c5d-e25c-41a3-ba2d-29f8ceb82ac1\"},\"php_version\":\"7.3\",\"site\":{\"id\":\"4d91f2dd-9743-421e-9bcf-20dcf0a195b2\"},\"environment\":\"production\",\"primary_domain\":\"jupiterone.com\",\"is_multisite\":false}]}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "content-length",
              "value": "398"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "date",
              "value": "Mon, 05 Jul 2021 11:52:24 GMT"
            },
            {
              "name": "x-amzn-requestid",
              "value": "10083c1d-d053-4818-bca8-ce9648e422f1"
            },
            {
              "name": "access-control-allow-origin",
              "value": "*"
            },
            {
              "name": "access-control-allow-headers",
              "value": "Content-Type,X-Amz-Date,Authorization,X-Api-Key"
            },
            {
              "name": "x-amz-apigw-id",
              "value": "B_rSyFFooAMFT6w="
            },
            {
              "name": "access-control-allow-methods",
              "value": "*"
            },
            {
              "name": "x-amzn-trace-id",
              "value": "Root=1-60e2f278-7ee905aa0ef9bc2315f60854"
            },
            {
              "name": "x-cache",
              "value": "Miss from cloudfront"
            },
            {
              "name": "via",
              "value": "1.1 52c6facb6d17a6f0790a93a219b39c3c.cloudfront.net (CloudFront)"
            },
            {
              "name": "x-amz-cf-pop",
              "value": "MNL50-C1"
            },
            {
              "name": "x-amz-cf-id",
              "value": "1uHPwgtqgafBXobIPn9h8qoyh51H4b2Dt8FSW3kuf8H2tjOmJtPLwQ=="
            }
          ],
          "headersSize": 601,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-07-05T11:52:24.008Z",
        "time": 645,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 645
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
