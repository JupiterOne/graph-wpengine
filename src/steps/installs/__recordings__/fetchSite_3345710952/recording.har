{
  "log": {
    "_recordingName": "fetchSite",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "4.3.0"
    },
    "entries": [
      {
        "_id": "b5a41916ddb8fd6f9a367486771ee186",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "api.wpengineapi.com"
            }
          ],
          "headersSize": 352,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "limit",
              "value": "10"
            },
            {
              "name": "offset",
              "value": "0"
            }
          ],
          "url": "https://api.wpengineapi.com/v1/accounts?limit=10&offset=0"
        },
        "response": {
          "bodySize": 112,
          "content": {
            "mimeType": "application/json",
            "size": 112,
            "text": "{\"previous\":null,\"next\":null,\"count\":1,\"results\":[{\"id\":\"77815c5d-e25c-41a3-ba2d-29f8ceb82ac1\",\"name\":\"j1inc\"}]}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "content-length",
              "value": "112"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "date",
              "value": "Mon, 21 Jun 2021 11:56:54 GMT"
            },
            {
              "name": "x-amzn-requestid",
              "value": "6fe32faa-cea4-402b-90a8-90e0f009a811"
            },
            {
              "name": "access-control-allow-origin",
              "value": "*"
            },
            {
              "name": "access-control-allow-headers",
              "value": "Content-Type,X-Amz-Date,Authorization,X-Api-Key"
            },
            {
              "name": "x-amz-apigw-id",
              "value": "BRi1BFgHoAMFQIw="
            },
            {
              "name": "access-control-allow-methods",
              "value": "*"
            },
            {
              "name": "x-amzn-trace-id",
              "value": "Root=1-60d07e86-41ca12b5004aa0b31221a0a5"
            },
            {
              "name": "x-cache",
              "value": "Miss from cloudfront"
            },
            {
              "name": "via",
              "value": "1.1 bd2a712c8280f0f57859c608168ce9f2.cloudfront.net (CloudFront)"
            },
            {
              "name": "x-amz-cf-pop",
              "value": "MNL50-C1"
            },
            {
              "name": "x-amz-cf-id",
              "value": "AgHRhi_dp2QkiMgwA6qj28zmKCegJiRqPpfoWO9MvY9kHgPBuZ2zQQ=="
            }
          ],
          "headersSize": 601,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-06-21T11:56:47.018Z",
        "time": 1119,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 1119
        }
      },
      {
        "_id": "97123c15bf93205301118bdc565d93ea",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "api.wpengineapi.com"
            }
          ],
          "headersSize": 349,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "limit",
              "value": "10"
            },
            {
              "name": "offset",
              "value": "0"
            }
          ],
          "url": "https://api.wpengineapi.com/v1/sites?limit=10&offset=0"
        },
        "response": {
          "bodySize": 354,
          "content": {
            "mimeType": "application/json",
            "size": 354,
            "text": "{\"previous\":null,\"next\":null,\"count\":1,\"results\":[{\"id\":\"4d91f2dd-9743-421e-9bcf-20dcf0a195b2\",\"name\":\"jupiterone\",\"account\":{\"id\":\"77815c5d-e25c-41a3-ba2d-29f8ceb82ac1\"},\"installs\":[{\"id\":\"30c47931-a91d-468a-a5dc-b9711f06c60b\",\"name\":\"jupiterone\",\"environment\":\"production\",\"cname\":\"jupiterone.wpengine.com\",\"php_version\":\"7.3\",\"is_multisite\":false}]}]}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "content-length",
              "value": "354"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "date",
              "value": "Mon, 21 Jun 2021 11:56:56 GMT"
            },
            {
              "name": "x-amzn-requestid",
              "value": "ec965d8b-75bc-414d-a215-e4414bb63e4f"
            },
            {
              "name": "access-control-allow-origin",
              "value": "*"
            },
            {
              "name": "access-control-allow-headers",
              "value": "Content-Type,X-Amz-Date,Authorization,X-Api-Key"
            },
            {
              "name": "x-amz-apigw-id",
              "value": "BRi1NGxhIAMF6HQ="
            },
            {
              "name": "access-control-allow-methods",
              "value": "*"
            },
            {
              "name": "x-amzn-trace-id",
              "value": "Root=1-60d07e87-1887dc8b33ebfcf14ed4b5db"
            },
            {
              "name": "x-cache",
              "value": "Miss from cloudfront"
            },
            {
              "name": "via",
              "value": "1.1 f92eab68beb1e6605042ec06f0941a64.cloudfront.net (CloudFront)"
            },
            {
              "name": "x-amz-cf-pop",
              "value": "MNL50-C1"
            },
            {
              "name": "x-amz-cf-id",
              "value": "W7_2e9K7cCGhAqOL5Q43pfHDKLstgQuEwTlPDZF4k1ooTTvJDfoVbw=="
            }
          ],
          "headersSize": 601,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-06-21T11:56:48.169Z",
        "time": 1188,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 1188
        }
      },
      {
        "_id": "7c0782aad0b9678323207b7733ae2209",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "api.wpengineapi.com"
            }
          ],
          "headersSize": 352,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "limit",
              "value": "10"
            },
            {
              "name": "offset",
              "value": "0"
            }
          ],
          "url": "https://api.wpengineapi.com/v1/installs?limit=10&offset=0"
        },
        "response": {
          "bodySize": 398,
          "content": {
            "mimeType": "application/json",
            "size": 398,
            "text": "{\"previous\":null,\"next\":null,\"count\":1,\"results\":[{\"id\":\"30c47931-a91d-468a-a5dc-b9711f06c60b\",\"name\":\"jupiterone\",\"cname\":\"jupiterone.wpengine.com\",\"stable_ips\":null,\"status\":\"active\",\"account\":{\"id\":\"77815c5d-e25c-41a3-ba2d-29f8ceb82ac1\"},\"php_version\":\"7.3\",\"site\":{\"id\":\"4d91f2dd-9743-421e-9bcf-20dcf0a195b2\"},\"environment\":\"production\",\"primary_domain\":\"jupiterone.com\",\"is_multisite\":false}]}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "content-length",
              "value": "398"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "date",
              "value": "Mon, 21 Jun 2021 11:56:57 GMT"
            },
            {
              "name": "x-amzn-requestid",
              "value": "ce598b95-c643-4ca9-a56b-9bc40ed8c073"
            },
            {
              "name": "access-control-allow-origin",
              "value": "*"
            },
            {
              "name": "access-control-allow-headers",
              "value": "Content-Type,X-Amz-Date,Authorization,X-Api-Key"
            },
            {
              "name": "x-amz-apigw-id",
              "value": "BRi1ZE0soAMFvIQ="
            },
            {
              "name": "access-control-allow-methods",
              "value": "*"
            },
            {
              "name": "x-amzn-trace-id",
              "value": "Root=1-60d07e88-32eeec243639094a4dc5c373"
            },
            {
              "name": "x-cache",
              "value": "Miss from cloudfront"
            },
            {
              "name": "via",
              "value": "1.1 5d93d2854733fe53280cae62f42d96de.cloudfront.net (CloudFront)"
            },
            {
              "name": "x-amz-cf-pop",
              "value": "MNL50-C1"
            },
            {
              "name": "x-amz-cf-id",
              "value": "FUKMQU-HuPnQ3ORm0MtMp4i1MTC8gwGWY-PCQRrlJrWavNnS6NBdPg=="
            }
          ],
          "headersSize": 601,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-06-21T11:56:49.369Z",
        "time": 1073,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 1073
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
