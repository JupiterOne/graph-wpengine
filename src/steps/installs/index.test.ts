import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';
import { IntegrationConfig } from '../../config';
import { integrationConfig } from '../../../test/config';
import { setupWPEngineRecording } from '../../../test/recording';
import { fetchInstalls } from '.';
import { fetchSites } from '../sites';
import { fetchAccounts } from '../accounts';
import { Relationships } from '../constants';

describe('#fetchInstalls', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupWPEngineRecording({
      directory: __dirname,
      name: 'fetchInstalls',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    await fetchAccounts(context);
    await fetchSites(context);
    await fetchInstalls(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    const sites = context.jobState.collectedEntities.filter((e) =>
      e._class.includes('Host'),
    );
    expect(sites.length).toBeGreaterThan(0);
    expect(sites).toMatchGraphObjectSchema({
      _class: ['Host'],
      schema: {
        additionalProperties: false,
        properties: {
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          _type: { const: 'wp_engine_site' },
          name: { type: 'string' },
        },
      },
    });

    const accounts = context.jobState.collectedEntities.filter((e) =>
      e._class.includes('Account'),
    );
    expect(accounts.length).toBeGreaterThan(0);
    expect(accounts).toMatchGraphObjectSchema({
      _class: ['Account'],
      schema: {
        additionalProperties: false,
        properties: {
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          _type: { const: 'wp_engine_account' },
          name: { type: 'string' },
        },
      },
    });

    const installs = context.jobState.collectedEntities.filter((e) =>
      e._class.includes('Application'),
    );
    expect(installs.length).toBeGreaterThan(0);
    expect(installs).toMatchGraphObjectSchema({
      _class: ['Application'],
      schema: {
        additionalProperties: false,
        properties: {
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          _type: { const: 'wp_engine_install' },
          name: { type: 'string' },
          phpVersion: { type: 'string' },
          status: { type: 'boolean' },
          cname: { type: 'string' },
          stableIps: { type: 'array', items: { type: 'string' } },
          environment: { type: 'string' },
          primaryDomain: { type: 'string' },
          isMultisite: { type: 'boolean' },
        },
      },
    });

    expect(
      context.jobState.collectedRelationships.filter(
        (e) => e._type === Relationships.ACCOUNT_HAS_INSTALL._type,
      ),
    ).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'HAS' },
          _type: {
            const: 'wp_engine_account_has_install',
          },
        },
      },
    });

    expect(
      context.jobState.collectedRelationships.filter(
        (e) => e._type === Relationships.SITE_HAS_INSTALL._type,
      ),
    ).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'HAS' },
          _type: {
            const: 'wp_engine_site_has_install',
          },
        },
      },
    });
  });
});
