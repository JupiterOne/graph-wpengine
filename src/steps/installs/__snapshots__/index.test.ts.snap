// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchInstalls should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "wp_engine_account:77815c5d-e25c-41a3-ba2d-29f8ceb82ac1",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "77815c5d-e25c-41a3-ba2d-29f8ceb82ac1",
            "name": "j1inc",
          },
        },
      ],
      "_type": "wp_engine_account",
      "createdOn": undefined,
      "displayName": "j1inc",
      "id": "77815c5d-e25c-41a3-ba2d-29f8ceb82ac1",
      "name": "j1inc",
    },
    Object {
      "_class": Array [
        "Host",
      ],
      "_key": "wp_engine_site:4d91f2dd-9743-421e-9bcf-20dcf0a195b2",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "account": Object {
              "id": "77815c5d-e25c-41a3-ba2d-29f8ceb82ac1",
            },
            "id": "4d91f2dd-9743-421e-9bcf-20dcf0a195b2",
            "installs": Array [
              Object {
                "cname": "jupiterone.wpengine.com",
                "environment": "production",
                "id": "30c47931-a91d-468a-a5dc-b9711f06c60b",
                "is_multisite": false,
                "name": "jupiterone",
                "php_version": "7.3",
              },
            ],
            "name": "jupiterone",
          },
        },
      ],
      "_type": "wp_engine_site",
      "createdOn": undefined,
      "displayName": "jupiterone",
      "id": "4d91f2dd-9743-421e-9bcf-20dcf0a195b2",
      "name": "jupiterone",
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "wp_engine_install:30c47931-a91d-468a-a5dc-b9711f06c60b",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "account": Object {
              "id": "77815c5d-e25c-41a3-ba2d-29f8ceb82ac1",
            },
            "cname": "jupiterone.wpengine.com",
            "environment": "production",
            "id": "30c47931-a91d-468a-a5dc-b9711f06c60b",
            "is_multisite": false,
            "name": "jupiterone",
            "php_version": "7.3",
            "primary_domain": "jupiterone.com",
            "site": Object {
              "id": "4d91f2dd-9743-421e-9bcf-20dcf0a195b2",
            },
            "stable_ips": null,
            "status": "active",
          },
        },
      ],
      "_type": "wp_engine_install",
      "active": true,
      "cname": "jupiterone.wpengine.com",
      "createdOn": undefined,
      "displayName": "jupiterone",
      "environment": "production",
      "id": "30c47931-a91d-468a-a5dc-b9711f06c60b",
      "isMultisite": false,
      "name": "jupiterone",
      "phpVersion": "7.3",
      "primaryDomain": "jupiterone.com",
      "stableIps": Array [],
      "status": true,
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "wp_engine_account:77815c5d-e25c-41a3-ba2d-29f8ceb82ac1",
      "_key": "wp_engine_account:77815c5d-e25c-41a3-ba2d-29f8ceb82ac1|has|wp_engine_site:4d91f2dd-9743-421e-9bcf-20dcf0a195b2",
      "_toEntityKey": "wp_engine_site:4d91f2dd-9743-421e-9bcf-20dcf0a195b2",
      "_type": "wp_engine_account_has_site",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "wp_engine_account:77815c5d-e25c-41a3-ba2d-29f8ceb82ac1",
      "_key": "wp_engine_account:77815c5d-e25c-41a3-ba2d-29f8ceb82ac1|has|wp_engine_install:30c47931-a91d-468a-a5dc-b9711f06c60b",
      "_toEntityKey": "wp_engine_install:30c47931-a91d-468a-a5dc-b9711f06c60b",
      "_type": "wp_engine_account_has_install",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "wp_engine_site:4d91f2dd-9743-421e-9bcf-20dcf0a195b2",
      "_key": "wp_engine_site:4d91f2dd-9743-421e-9bcf-20dcf0a195b2|has|wp_engine_install:30c47931-a91d-468a-a5dc-b9711f06c60b",
      "_toEntityKey": "wp_engine_install:30c47931-a91d-468a-a5dc-b9711f06c60b",
      "_type": "wp_engine_site_has_install",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "wp_engine_account",
    "wp_engine_site",
    "wp_engine_account_has_site",
    "wp_engine_install",
    "wp_engine_account_has_install",
    "wp_engine_site_has_install",
  ],
  "numCollectedEntities": 3,
  "numCollectedRelationships": 3,
}
`;
